Tests/NIOHTTP1Tests/HTTPDecoderLengthTest.swift

```diff
@@ -0,0 +1,519 @@
+        try assertResponseTransferEncodingHasBodyTerminatedByEOF(transferEncodingHeader: "chunked, gzip", eofMechanism: .halfClosure)
+    }
+
+    func testMultipleTEWithChunkedInTheMiddleHasEOFBodyOnResponseWithHalfClosure() throws {
+        // Here http_parser is right, and this is EOF terminated.
+        try assertResponseTransferEncodingHasBodyTerminatedByEOF(transferEncodingHeader: "gzip, chunked, deflate", eofMechanism: .halfClosure)
+    }
+
+    func testRequestWithTEAndContentLengthErrors() throws {
+        XCTAssertNoThrow(try channel.pipeline.add(handler: HTTPRequestDecoder()).wait())
+
+        // Send a GET with the invalid headers.
+        do {
+            try channel.writeInbound(ByteBuffer(string: "POST / HTTP/1.1\r\nTransfer-Encoding: chunked\r\nContent-Length: 4\r\n\r\n"))
+            XCTFail("Did not throw")
+        } catch HTTPParserError.unexpectedContentLength {
+            // ok
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+
+        // Must spin the loop.
+        (channel.eventLoop as! EmbeddedEventLoop).run()
+    }
+
+    func testResponseWithTEAndContentLengthErrors() throws {
+        XCTAssertNoThrow(try channel.pipeline.add(handler: HTTPRequestEncoder()).wait())
+        XCTAssertNoThrow(try channel.pipeline.add(handler: HTTPResponseDecoder()).wait())
+
+        // Prime the decoder with a request.
+        XCTAssertTrue(try channel.writeOutbound(HTTPClientRequestPart.head(HTTPRequestHead(version: .init(major: 1, minor: 1),
+                                                                                           method: .GET,
+                                                                                           uri: "/"))))
+
+        // Send a 200 OK with the invalid headers.
+        do {
+            try channel.writeInbound(ByteBuffer(string: "HTTP/1.1 200 OK\r\nTransfer-Encoding: chunked\r\nContent-Length: 4\r\n\r\n"))
+            XCTFail("Did not throw")
+        } catch HTTPParserError.unexpectedContentLength {
+            // ok
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+
+        // Must spin the loop.
+        (channel.eventLoop as! EmbeddedEventLoop).run()
+    }
+
+    private func assertRequestWithInvalidCLErrors(contentLengthField: String) throws {
+        XCTAssertNoThrow(try channel.pipeline.add(handler: HTTPRequestDecoder()).wait())
+
+        // Send a GET with the invalid headers.
+        do {
+            try channel.writeInbound(ByteBuffer(string: "POST / HTTP/1.1\r\nContent-Length: \(contentLengthField)\r\n\r\n"))
+            XCTFail("Did not throw")
+        } catch HTTPParserError.invalidContentLength {
+            // ok
+        } catch HTTPParserError.unexpectedContentLength {
+            // also ok
+        } catch {
+            XCTFail("Unexpected error: \(error)")
+        }
+
+        // Must spin the loop.
+        (channel.eventLoop as! EmbeddedEventLoop).run()
```

Comment:
assert the channel is not native anymore.

Reply:
I think we should actually assert the return value of channel.finish()

Reply:
Nope, we can't do that here because `channel.finish()` will throw if the channel is already closed.